kolibri/auth/permissions.py

```diff
@@ -0,0 +1,349 @@
+"""
+This module defines the base classes for Kolibri's class-based Permissions system. Other apps can import the classes
+from this module in their own "permissions.py" module, and extend them, and then apply them to their own models.
+"""
+
+
+#####################################################################################################################
+### This first section consists of the base permissions classes: the ones likely to be imported and extended.     ###
+#####################################################################################################################
+
+class BasePermissions(object):
+    """
+    Base Permission class from which all other Permission classes should inherit.
+    By default, it denies permissions for everything. To be more permissive, override its methods.
+    Note that if you don't need to distinguish between create, update, and delete permissions, you can
+    simply override `_user_can_write_object` to define behavior for them in one place.
+    """
+
+    def user_can_create_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to create the provided <obj>.
+        Note that the object may not yet have been saved to the database (as this may be a pre-save check)."""
+        return self._user_can_write_object(user, obj)
+
+    def user_can_read_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to read the provided <obj>."""
+        return False
+
+    def user_can_update_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to update the provided <obj>."""
+        return self._user_can_write_object(user, obj)
+
+    def user_can_delete_object(self, user, obj):
+        """Returns True if this permission class grants <user> permission to delete the provided <obj>."""
+        return self._user_can_write_object(user, obj)
+
+    def _user_can_write_object(self, user, obj):
```

Comment:
I could see advantages in that, yeah -- for example, then `user_can_write_object` could be called directly, externally. Except that's also a disadvantage -- since `user_can_write_object` wouldn't always be there, we couldn't ever actually safely call it in an agnostic way, since the method wouldn't always be defined. And even if it were always there, we'd never want to call it externally since we may have overridden the more specific permission method.

If it feels awkward having a method on the base class that won't always get used, perhaps we just leave it out entirely, have everything return `False` by default, and allow subclasses to define their own DRY way of lumping multiple permission types together, as appropriate?